### ISSUE TITLE
Wrong warning for Template Haskell with Type Application

### ISSUE THREAD

**Issue by odr** [ndmitchell/hlint#491]
In HLint v2.1.3 I've got warning:

> Redundant bracket
Found: ($x)
Why not: $x

But there is a type application @($x) and parenthesis are needed.

**Comment by odr** [comment]
Well, when I write
```haskell
{-# LANGUAGE TemplateHaskell  #-}
{-# LANGUAGE TypeApplications #-}
module Test where
import           Language.Haskell.TH

f x = [e| read @($(conT x))|]
g x = [e| read @($x)|]
```
I got messages: 

> Redundant bracket
> Found:   ($( conT x ))
> Why not:   $( conT x )
> 

> Redundant bracket
> Found:   ($x)
> Why not:   $x
>

**Comment by ndmitchell** [comment]
I tried following the hints and got:

```haskell
f x = [e| read @ $(conT x) |]
g x = [e| read @ $x |]
```

So it seems the brackets really are redundant?

**Comment by odr** [comment]
Well, Neil, you are right, sorry.

I can't find how I got problem with it. I've found now that in some cases space between @ and $ is needed and sometimes is optional. But it is not a hlint's problem.

Thank you!

**Comment by ndmitchell** [comment]
Thanks for letting me know! I think you need a space if the first character is a symbol, e.g. `$`, since `@$` forms a lexeme so gets parsed differently.
